### PR TITLE
[Part 2] Calendar to GCS Operator

### PR DESCRIPTION
**Pull request by rsg17**
related: #8471

This PR adds a Google Calendar to GCS Operator to write Google Calendar Events to GCS.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by rsg17** [comment]
I would like some help on how to do system tests. I have created the example dag and system test file; but I am not sure how to test it.

I read through this [guide](https://github.com/apache/airflow/blob/main/TESTING.rst#airflow-system-tests). I think I need make some changes to variables.env file; but I am not sure what these changes should be.

Also, I am not sure about the --forward-credentials option for [running system tests](https://github.com/apache/airflow/blob/main/TESTING.rst#the-typical-system-test-session). Which credentials will it forward? Do I need to put the credentials somewhere?

**Comment by turbaszek** [comment]
> 
> I read through this [guide](https://github.com/apache/airflow/blob/main/TESTING.rst#airflow-system-tests). I think I need make some changes to variables.env file; but I am not sure what these changes should be.

From your example DAG it looks like you need to set up two variables:
- GCP_GCS_BUCKET
- CALENDAR_ID

Apart from that you will need to add a credentials key because you are using it in test:
```py
@pytest.mark.credential_file(GCP_GCS_KEY)
```
The file should be named `gcp_gcs.json` and put according to the guid you linked:

> If your system tests need some credential files to be available for an authentication with external systems, make sure to keep these credentials in the files/airflow-breeze-config/keys directory. Mark your tests with @pytest.mark.credential_file(<FILE>) so that they are skipped if such a credential file is not there. The tests should read the right credentials and authenticate them on their own. The credentials are read in Breeze from the /files directory. The local "files" folder is mounted to the "/files" folder in Breeze.


> Also, I am not sure about the --forward-credentials option for [running system tests](https://github.com/apache/airflow/blob/main/TESTING.rst#the-typical-system-test-session). Which credentials will it forward? Do I need to put the credentials somewhere?

You need to use credentials to some GCP project. As far as I know we don't have airflow project that can be used for this purpose (cc @potiuk to confirm).

**Comment by rsg17** [comment]
Thank you!

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py:28]: There are several input parameters to this operator that are missing from the docstring. Would you mind adding them? The docstring is directly used to build the operator's API documentation in Airflow.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]:
```suggestion
        show_deleted: bool = False,
        show_hidden_invitation: bool = False,
```
The `Optional` typing implies that this value _could_ be `None`. It doesn't seem like this is the case.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: It would be great if this parameter name was something more verbose. I don't think users would have a practical idea of what this is related to by its name alone.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: This looks to be unused in the method.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: This looks like it could be moved to the `_upload_data()` method where it is solely used rather than instantiating it in `execute()` and passing the object around (or making it an instance attribute).

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Does this need to be a list? The `_upload_data()` method only returns a single file name and the method isn't called multiple times when the operator is executed.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: This feels redundant. The operator is effectively pushing the same value twice as an `XCom` but with different keys: the direct `xcom_push()` will push with an `XCom` key of "destination_objects" while the return would push with a key of "return_value" (assuming users don't set `do_xcom_push=False`).

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Is this `TODO` still needed? The associated doc looks relatively complete.

You can always mark PRs as a draft first if you still have some WIP actions you'd like to do before you'd like a more detailed review.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]:
```suggestion
    :param destination_bucket: The destination Google Cloud Storage bucket where the
        report should be written to. (templated)
    :type destination_bucket: str
    :param destination_path: The Google Cloud Storage URI array for the object created by the operator.
        For example: ``path/to/my/files``.
```
Small nit for consistency.

File: docs/apache-airflow-providers-google/operators/transfer/calendar_to_gcs.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-google/operators/transfer/calendar_to_gcs.rst]:
```suggestion
Google Calendar to Google Cloud Storage Transfer Operators
==========================================================
```
The underline needs to be at least as long as that which is being underlined.

File: docs/apache-airflow-providers-google/operators/transfer/calendar_to_gcs.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-google/operators/transfer/calendar_to_gcs.rst]:
```suggestion
Upload data from Google Calendar to GCS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by josh-fell — COMMENTED** [review]
Looks pretty good! Just a few suggestions.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py:28]: Thank you for your review @josh-fell!

Will add it.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: The parameter is Optional and the default value is False.

I think either suggestion of making in `None` or not `Optional` will work. I will test it out.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Yes - will rename it to `free_text_query` or something similar based on https://developers.google.com/calendar/api/v3/reference/events/list

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Actually I marked it as `TODO` because I think it needs the doc url.. which I assumed would be created after I merged this PR.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py:28]: Added

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Removed `TODO`

**Comment by rsg17** [comment]
> > I read through this [guide](https://github.com/apache/airflow/blob/main/TESTING.rst#airflow-system-tests). I think I need make some changes to variables.env file; but I am not sure what these changes should be.
> 
> From your example DAG it looks like you need to set up two variables:
> 
> * GCP_GCS_BUCKET
> * CALENDAR_ID
> 
> Apart from that you will need to add a credentials key because you are using it in test:
> 
> ```python
> @pytest.mark.credential_file(GCP_GCS_KEY)
> ```
> 
> The file should be named `gcp_gcs.json` and put according to the guide you linked:
> 
> > If your system tests need some credential files to be available for an authentication with external systems, make sure to keep these credentials in the files/airflow-breeze-config/keys directory. Mark your tests with @pytest.mark.credential_file() so that they are skipped if such a credential file is not there. The tests should read the right credentials and authenticate them on their own. The credentials are read in Breeze from the /files directory. The local "files" folder is mounted to the "/files" folder in Breeze.
> 
> Regarding this question:
> 
> > Also, I am not sure about the --forward-credentials option for [running system tests](https://github.com/apache/airflow/blob/main/TESTING.rst#the-typical-system-test-session). Which credentials will it forward? Do I need to put the credentials somewhere?
> 
> You need to use credentials to some GCP project. As far as I know we don't have airflow project that can be used for this purpose (cc @potiuk to confirm).

Thank you for all your help @turbaszek. 
I am running into an authentication error. I don't see this error when I try to connect to gcs directly. Could you help take a look?

This gives the error location:
`___________________________________ GoogleCalendarToGCSExampleDagsSystemTest.test_run_example_dag_function ____________________________________
/usr/local/lib/python3.7/contextlib.py:73: in inner
    with self._recreate_cm():
/usr/local/lib/python3.7/contextlib.py:112: in __enter__
    return next(self.gen)
tests/test_utils/gcp_system_helpers.py:108: in provide_gcp_context
    f"--key-file={key_file_path}",
tests/test_utils/logging_command_executor.py:92: in execute_cmd
    env=env,
/usr/local/lib/python3.7/subprocess.py:800: in __init__
    restore_signals, start_new_session)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <subprocess.Popen object at 0x4065c8b250>
args = ['gcloud', 'auth', 'activate-service-account', '--key-file=/files/airflow-breeze-config/keys/gcp_gcs.json'], executable = b'gcloud'
preexec_fn = None, close_fds = True, pass_fds = (), cwd = None, env = None, startupinfo = None, creationflags = 0, shell = False, p2cread = -1
p2cwrite = -1, c2pread = 18, c2pwrite = 19, errread = 20, errwrite = 21, restore_signals = True, start_new_session = False
`

Here is the error:
`E               FileNotFoundError: [Errno 2] No such file or directory: 'gcloud': 'gcloud'`

I tried running `gcloud auth activate-service-account` with the same key and same service account from my laptop - that worked. I was also able to upload a file to GCS using the same key: https://gist.github.com/rsg17/b582414f630c2ac38323b87213b270d5

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py:28]: Based on the newly merged #20951, can you remove the `:type:` Sphinx directives from the operator docstring? The API documentation is able to infer from the type annotations of parameters now rather than relying on the docstring.

File: airflow/providers/google/cloud/transfers/calendar_to_gcs.py
Comment (josh-fell) [inline airflow/providers/google/cloud/transfers/calendar_to_gcs.py]:
```suggestion
        :ref:`howto/operator:GoogleCalendarToGCSOperator`
```

This should help with the Build Docs errors.

Comment (rsg17) [reply airflow/providers/google/cloud/transfers/calendar_to_gcs.py]: Thank you!

**Comment by rsg17** [comment]
@turbaszek 
Was able to resolve the issues with system tests and could verify that the events I created on a test calendar were available as JSON file on GCS.
Screen Shot 2022-01-31 at 10 32 38 PM

**Review by eladkal — APPROVED** [review]
LGTM
@rsg17 can you please rebase

If no further comments i'll merge after

**Review by josh-fell — APPROVED** [review]

**Comment by rsg17** [comment]
> LGTM @rsg17 can you please rebase
> 
> If no further comments i'll merge after

Rebased. Thank you!

**Pull request merged**
